inference-engine/src/auto_plugin/auto_plugin.cpp

```diff
@@ -0,0 +1,206 @@
+        try {
+            if (value.deviceName != _pluginName) {
+                auto deviceQr = GetCore()->QueryNetwork(network, value.deviceName, value.config);
+                std::unordered_set<std::string> deviceSupportedLayers;
+                for (auto &&layerQr : deviceQr.supportedLayersMap) {
+                    deviceSupportedLayers.emplace(layerQr.first);
+                }
+                supportedLayers = supportedLayers.empty()
+                                ? deviceSupportedLayers : (deviceSupportedLayers.empty()
+                                ? supportedLayers : IE::details::Intersection(
+                                     supportedLayers, deviceSupportedLayers));
+                supportedDevices.insert(value.deviceName);
+            }
+        } catch (...) {
+        }
+    }
+
+    if (supportedDevices.empty()) {
+        IE_THROW() << "Please, check environment due to no supported devices can be used";
+    }
+
+    for (auto&& supportedLayer : supportedLayers) {
+        queryResult.supportedLayersMap[supportedLayer] = GetName();
+    }
+    return queryResult;
+}
+
+IE::Parameter AutoInferencePlugin::GetConfig(const std::string& name,
+                                             const std::map<std::string, IE::Parameter> & options) const {
+    auto it = _config.find(name);
+    if (it == _config.end()) {
+        IE_THROW() << "Unsupported config key: " << name;
+    } else {
+        return { it->second };
+    }
+}
+
+void AutoInferencePlugin::SetConfig(const ConfigType& config) {
+    for (auto && kvp : config) {
+        _config[kvp.first] = kvp.second;
+    }
+}
+
+IE::Parameter AutoInferencePlugin::GetMetric(const std::string& name,
+                                             const std::map<std::string, IE::Parameter> & options) const {
+    if (name == METRIC_KEY(SUPPORTED_METRICS)) {
+        std::vector<std::string> metrics;
+        metrics.emplace_back(METRIC_KEY(AVAILABLE_DEVICES));
+        metrics.emplace_back(METRIC_KEY(SUPPORTED_METRICS));
+        metrics.emplace_back(METRIC_KEY(FULL_DEVICE_NAME));
+        metrics.emplace_back(METRIC_KEY(SUPPORTED_CONFIG_KEYS));
+        metrics.emplace_back(METRIC_KEY(OPTIMIZATION_CAPABILITIES));
+        IE_SET_METRIC_RETURN(SUPPORTED_METRICS, metrics);
+    } else if (name == METRIC_KEY(AVAILABLE_DEVICES)) {
+        // FIXME: available devices should get from thirdparty
+        std::vector<std::string> availableDevices = { "" };
+        IE_SET_METRIC_RETURN(AVAILABLE_DEVICES, availableDevices);
+    } else if (name == METRIC_KEY(FULL_DEVICE_NAME)) {
+        std::string device_name = { "AUTO" };
+        IE_SET_METRIC_RETURN(FULL_DEVICE_NAME, device_name);
+    } else if (name == METRIC_KEY(SUPPORTED_CONFIG_KEYS)) {
+        std::vector<std::string> configKeys;
+        IE_SET_METRIC_RETURN(SUPPORTED_CONFIG_KEYS, configKeys);
+    } else if (name == METRIC_KEY(OPTIMIZATION_CAPABILITIES)) {
+        std::vector<std::string> capabilities = { "" };
```

Comment:
`""` is not a valid optimization capability. W discussed that AUTO should combine all optimizations capabilities rom device plugins.

Reply:
well, I'd like to return empty for this kind of capabilities now, this PR contains too much features, we could improve this in another PR, that will be more easy and only few code change.